File: crates/kaya/examples/todos.ml
(* The todos scene from OCaml, on the let* surface: records and field
   projection. The descriptor is the schema — the honest floor a future
   ppx deriver will generate — the template binds each field to its own
   widget through typed field tokens, and toggling a row sends one
   field's delta through update_field: the title never travels.

   Build like milestone2.ml, then run with KAYA_SELFTEST=todos. *)

open Kaya_wire
open Kaya_app

(* The record and its descriptor: schema, conversions, and field
   tokens, adjacent by discipline until [@@deriving kaya] deletes the
   obligation. *)
type todo = { title : string; done_ : bool }

let todo_type : todo record_type =
  {
    rt_schema = [ value_str; value_bool ];
    rt_to_values = (fun t -> [ Str t.title; Bool t.done_ ]);
    rt_of_values =
      (function
      | [ Str title; Bool done_ ] -> { title; done_ }
      | _ -> invalid_arg "kaya: todo fields out of order");
  }

let field_title : (todo, string) field = str_field 0
let field_done : (todo, bool) field = bool_field 1

let () =
  let app = Kaya_app.create () in

  let items_left_text todos =
    let* entries = record_items todos in
    let n = List.length (List.filter (fun (_, t) -> not t.done_) entries) in
    io (fun () -> if n = 1 then "1 item left" else Printf.sprintf "%d items left" n)
  in

  let items_left, field, add, todos, check =
    build app
      (let* items_left = signal (Str "0 items left") in

       let* column = widget kind_column in
       let* field = widget kind_entry in
       let* add = widget kind_button in
       let* () = set_text add "Add" in
       let* status = widget kind_label in
       let* () = bind_text status items_left in

       let* todos = collection_of todo_type in
       let* todo_list, check =
         for_each (record_handle todos)
           Tpl.(
             let* row = widget kind_row in
             let* check = widget kind_checkbox in
             let* () = bind_checked_field check field_done in
             let* title = widget kind_label in
             let* () = bind_text_field title field_title in
             let* () = add_child row check in
             let+ () = add_child row title in
             check)
       in

       let* () = add_child column field in
       let* () = add_child column add in
       let* () = add_child column status in
       let* () = add_child column todo_list in
       let+ () = mount column in
       (items_left, field, add, todos, check))
  in

  (* The fold: widget-owned state arrives as occurrences; the app's
     copy is this ref, not a widget read. *)
  let draft = ref "" in
  let next_key = ref 0 in
  on_change app field (fun text -> io (fun () -> draft := text));
  on_click app add
    (let* key = io (fun () -> incr next_key; Printf.sprintf "t%d" !next_key) in
     let* () = insert_record todos (Str key) { title = !draft; done_ = false } in
     let* status = items_left_text todos in
     write items_left (Str status));
  on_toggle_node app check (fun keys checked ->
      (* One field's delta: the title never travels. *)
      let* () = update_field todos (List.hd keys) field_done checked in
      let* status = items_left_text todos in
      write items_left (Str status));

  exit (run app)

File: crates/kaya/examples/todos.go
// The todos scene from Go: records and field projection. The struct is
// the schema — CollectionOf reflects it once at declaration — the
// template binds each field to its own widget through typed field
// tokens, and toggling a row sends one field's delta through
// UpdateField: the title never travels.
//
// Build the library first (cargo build), then, from the repo root:
//
//	KAYA_SELFTEST=todos go run crates/kaya/examples/todos.go
package main

import (
	"fmt"
	"os"
	"runtime"

	kaya "dev.kaya/bindings/go"
)

func init() {
	// The core must own the process main thread.
	runtime.LockOSThread()
}

// Todo is the record type and, by reflection, the schema.
type Todo struct {
	Title string
	Done  bool
}

// The field tokens, checked against the struct at startup.
var (
	fieldTitle = kaya.FieldOf[Todo, string]("Title")
	fieldDone  = kaya.FieldOf[Todo, bool]("Done")
)

func main() {
	app := kaya.NewApp()

	var (
		itemsLeft kaya.Signal
		field     kaya.Widget
		add       kaya.Widget
		todos     kaya.RecordCollection[Todo]
		check     kaya.Node
	)

	itemsLeftText := func(tx *kaya.Tx) string {
		n := 0
		for _, e := range todos.Items(tx) {
			if !e.Value.Done {
				n++
			}
		}
		if n == 1 {
			return "1 item left"
		}
		return fmt.Sprintf("%d items left", n)
	}

	app.Build(func(tx *kaya.Tx) {
		itemsLeft = tx.Signal("0 items left")

		column := tx.Widget(kaya.KindColumn)
		field = tx.Widget(kaya.KindEntry)
		add = tx.Widget(kaya.KindButton)
		tx.SetText(add, "Add")
		status := tx.Widget(kaya.KindLabel)
		tx.BindText(status, itemsLeft)

		todos = kaya.CollectionOf[Todo](tx)
		todoList := tx.ForEach(todos.Collection, func(t *kaya.Tpl) {
			row := t.Widget(kaya.KindRow)
			check = t.Widget(kaya.KindCheckbox)
			t.BindCheckedField(check, 0, fieldDone)
			title := t.Widget(kaya.KindLabel)
			t.BindTextField(title, 0, fieldTitle)
			t.AddChild(row, check)
			t.AddChild(row, title)
		})

		tx.AddChild(column, field)
		tx.AddChild(column, add)
		tx.AddChild(column, status)
		tx.AddChild(column, todoList)
		tx.Mount(column)
	})

	// The fold: widget-owned state arrives as occurrences; the app's
	// copy is this variable, not a widget read.
	draft := ""
	nextKey := 0
	app.OnChange(field, func(tx *kaya.Tx, text string) {
		draft = text
	})
	app.OnClick(add, func(tx *kaya.Tx) {
		nextKey++
		todos.Insert(tx, fmt.Sprintf("t%d", nextKey), Todo{Title: draft})
		tx.Write(itemsLeft, itemsLeftText(tx))
	})
	app.OnToggleNode(check, func(tx *kaya.Tx, keys []any, checked bool) {
		// One field's delta: the title never travels.
		kaya.UpdateField(tx, todos, keys[0], fieldDone, checked)
		tx.Write(itemsLeft, itemsLeftText(tx))
	})

	os.Exit(app.Run())
}
